src/Conversion/ONNXToKrnl/Math/Elementwise.cpp

```diff
@@ -1735,29 +1748,20 @@ bool OpFusionHelper::isControlFlowValidForFusion(
 // assumed the canonicalization has hoisted all constant to the beginning of the
 // function by fold function.
 bool OpFusionHelper::areInputsValidForFusion(
-    Operation *useOp, Operation *defOp) {
+    Operation *useOp, Operation *defOp, DimAnalysis *dimAnalysis) {
   // Elementwise unary operation is always fusible
   if (useOp->getOperands().size() == 1)
     return true;
 
-  // To fuse Elementwise op with more one operands with the producer,
-  // the shape of the output the user Op has to have the same size
-  // output as that of the producer Op. Here dimension expansion with size
-  // 1 is allowed. Refer to hasNoBroadcast() definition.
-  // ToFix: This PR simply check static shape and does not use symbolic
-  // shape analysis and BroadcastShapeHelper
-  // Some discussion can be found at
-  // https://github.com/onnx/onnx-mlir/issues/2199
-
-  if (!hasStaticShape(defOp->getResults()[0].getType()))
-    return false;
-
-  ArrayRef<int64_t> defShape = getShape(defOp->getResults()[0].getType());
-  ArrayRef<int64_t> useShape = getShape(useOp->getResults()[0].getType());
+  Type defOutputType = defOp->getResultTypes()[0];
+  Type useOutputType = useOp->getResultTypes()[0];
+  ArrayRef<int64_t> defShape = getShape(defOutputType);
+  ArrayRef<int64_t> useShape = getShape(useOutputType);
   if (defShape != useShape) {
```

Comment:
This condition seems to be stronger than the code that starts in line 1786:
```
  if (getRank(defOutputType) != getRank(useOutputType))
     return false;

   if (dimAnalysis) {
     if (!dimAnalysis->sameShape(defOp->getResult(0), useOp->getResult(0)))
       return false;
   } else {
     // If there is no dimAnalysis, check the simplest case.
     // Static and the same shape
     if (!hasStaticShape(useOutputType))
       return false;

     ArrayRef<int64_t> inputShape = getShape(useOutputType);
     if (inputShape != defShape)
       return false;
```

Reply:
I assume that the comparison of defShape and useShape will treat the dynamic dimension as the same. The later code use DimAnalysis to analyze the dynamic dimension. Will check it.

Reply:
I tested with an example. My assumption is correct.